tmt/steps/prepare/distgit.py

```diff
@@ -0,0 +1,138 @@
+import dataclasses
+from typing import TYPE_CHECKING, Any, Optional
+
+import tmt
+import tmt.log
+import tmt.steps
+import tmt.steps.prepare
+import tmt.utils
+from tmt.steps.provision import Guest
+from tmt.utils import Command, Path, field
+
+if TYPE_CHECKING:
+    import tmt.steps.prepare.install
+
+PREPARE_WRAPPER_FILENAME = 'tmt-prepare-wrapper.sh'
+
+
+@dataclasses.dataclass
+class DistGitData(tmt.steps.prepare.PrepareStepData):
+    source_dir: Optional[Path] = field(
+        default=None,
+        option='--source-dir',
+        normalize=tmt.utils.normalize_path,
+        exporter=lambda value: str(value) if isinstance(value, Path) else None)
+    phase_name: str = field(default_factory=str, option='--phase-name')
+    order: int = 60
+    install_builddeps: bool = field(
+        default=False,
+        option="--install-builddeps",
+        is_flag=True,
+        help="Install package build dependencies",
+        )
+    require: list['tmt.base.DependencySimple'] = field(
+        default_factory=list,
+        option="--require",
+        metavar='PACKAGE',
+        multiple=True,
+        help='Additional required package to be present before sources are prepared',
+        # *simple* requirements only
+        normalize=lambda key_address, value, logger: tmt.base.assert_simple_dependencies(
+            tmt.base.normalize_require(key_address, value, logger),
+            "'require' can be simple packages only",
+            logger),
+        serialize=lambda packages: [package.to_spec() for package in packages],
+        unserialize=lambda serialized: [
+            tmt.base.DependencySimple.from_spec(package)
+            for package in serialized
+            ]
+        )
+
+
+@tmt.steps.provides_method('distgit')
+class PrepareDistGit(tmt.steps.prepare.PreparePlugin[DistGitData]):
+    """
+    Companion to the discover-dist-git
+
+    """
+
+    _data_class = DistGitData
+
+    def __init__(
+            self,
+            *,
+            discover: Optional['tmt.steps.discover.DiscoverPlugin'] = None,
```

Comment:
`DiscoverPlugin` is a generic plugin, and expects a type of plugin's "step data" container, e.g. `DiscoverFmfStepData` or `DiscoverShellData`. In your plugin the actual type does not matter, so, IIUIC, you have these options:

* "distgit plugin might be subclassed or instantiated with different discover plugins, and their step data type is important because this plugin accesses `self.discover.data`" - in this case, you would use `DiscoverStepDataT` which represents "discover step data-like" types. but since the distgit plugin does not work this way, mypy would probably complain that this type variable is left without an actual value. So I wouldn't go this way.
* "distgit plugin accepts discover plugin instance, and its step data type is not important" - `tmt.steps.discover.DiscoverPlugin[Any]` should do. Very open.
* and somewhere between these two you have "distgit plugin accepts discover plugin instance, and its step data type is not important - but step data of discover plugins  can only be an instance of `tmt.steps.discover.DiscoverStepData` or one of its subclasses". So it'd be more strict than the second option, without dealign with actual types in various plugins, `tmt.steps.discover.DiscoverPlugin[tmt.steps.discover.DiscoverStepData]` would be my choice. But mypy has often issues with checking this construct, so you may end up with `Any` which is not wrong, but a bit more open.

Reply:
fixed in [486c9e6](https://github.com/teemtee/tmt/pull/2641/commits/486c9e67abca28e4e923010376b7008a562260cc). Thanks a lot